### ISSUE TITLE
Specs should run in random order by default

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#12540]
The spec module has support for random order of execution (added in #8310). But by default, it runs examples in literal order ("--order=default").
You only get random order by specifying `--order=random` explicitly.

Random order of execution is a good default behaviour. It ensures that individual examples are not order-dependent in a way that example B only succeeds if example A was run before it.

So we should make it the default behaviour for spec execution.
This has been requested before (https://github.com/crystal-lang/crystal/issues/8289#issuecomment-541363468, https://github.com/crystal-lang/crystal/issues/8289#issuecomment-539391528) and I think it's a good cause to address in this separate issue.

This change *can* lead to broken specs. Although I would argue that's good. It shows that the specs are broken.

**Comment by mattrberry** [comment]
I agree that random by default has its benefits. Before making the change, though, it'd be nice to add a feature that lets you rerun specs in the order in which they failed.

For example, if a spec fails, it'd be great if a seed is printed to the console and that seed could be fed back to `crystal spec` to rerun the tests in the same order.

Even if random doesn't become the default, this would help in debugging flaky tests

**Comment by asterite** [comment]
I was going to suggest exactly that!

But... apparently that's already the case:


![image](https://user-images.githubusercontent.com/209371/193067676-5c489cbb-8ff5-4071-9bd8-65882a77a4c4.png)

**Comment by Fryguy** [comment]
When I originally wrote the randomization, I wanted random as the new default, but we didn't want to break anything with such a new feature.  However, now that it's been stable for a long time, I agree with making random the default.  Users can easily switch back to `--order=default` if they hit any issues they can't immediately fix.

**Comment by asterite** [comment]
I personally hate running spec in random order by default, but that's just me. When I get failures I find it pretty annoying to see the failures not in an ordered way... but now that I think about it, maybe we can show the failures in the usual order, even if they ran randomly 🤔

**Comment by straight-shoota** [comment]
> but now that I think about it, maybe we can show the failures in the usual order, even if they ran randomly 🤔

I don't really think that's a good idea. I imagine it would be wildly confusing.
And how would you be able to figure out which specs' side effects have an influence on a subsequent spec, when you don't see the actual order of execution?

**Comment by straight-shoota** [comment]
But I see your point @asterite. During development (especially when that's test driven) it can be helpful to execute specs in fixed, lexical order. This makes it easier to go through the failures top to bottom, which usually corresponds to the increasing complexity of the test cases. Tests are often structured in a way that more basic features are tested first, and then it goes into more details and edge cases. Fixing the simple ones first makes a lot of sense, and such changes often already influence more complex tests.
A fixed order also makes it easier to concentrate on specific tests and how they evolve over time.

So I guess that makes two use cases which demand different default behaviour. Let's figure out how to best tell them apart.

**Comment by asterite** [comment]
Just my opinion: a user wants predictable order, CI wants random order. So the default should be predictable order. But I'm fine either way.
